### DISCUSSION TITLE
Pandoc WASM

### DISCUSSION THREAD

**Discussion by tarleb** [jgm/pandoc#10387] [Show and tell]
For those interested: @TerrorJack successfully compiled pandoc to WASM via ghc-wasm-meta.

- Livedemo: https://tweag.github.io/pandoc-wasm/
- Repo: https://github.com/tweag/pandoc-wasm

Fantastic work!

**Comment by fiapps** [comment]
It's too bad Lua filters don't work yet. This makes it less useful.

I experimented with it briefly using the iOS app a-Shell. I copied the URL of `pandoc.wasm` from the README and downloaded it with `curl`. When I run it, it hangs, but if I press CTRL-C, it produces the expected output. The hang isn't due to an attempt to read from stdin, because CTRL-D has no effect. Probably, it needs to be build with a-Shell's [custom WASI-SDK](https://github.com/holzschu/wasi-sdk), and perhaps some of TerrorJack's customization needs to be undone. Since this is not `wasmtime` but another platform, it may be that Lua filters can be made to work. However, I don't expect to dedicate time to this in the near future. Since I have no experience with Haskell, I imagine I'd have a hard time making a custom build of pandoc.

**Comment by jgm** [comment]
@TerrorJack this is amazing!

I'm wondering whether we should explore modifying the main pandoc repo with a wasm build configuration, so that a separate fork is not needed -- perhaps with CPP conditionals.  Do you think this would be worthwhile?

**Reply by jgm** [reply]
Looking at this, I see
```hs
#if !defined(wasm32_HOST_ARCH)
                 | PandocHttpError Text HttpException
#endif
```
which would make the API dependent on the build configuration.  That's not desirable.  But we could contemplate changing the API so that PandocHttpError doesn't use HttpException, I suppose.

**Reply by TerrorJack** [reply]
> Do you think this would be worthwhile?

Absolutely! I was hesitant to open a PR because more CPP is more maintenance burden, but if you're open to accepting this upstream, I can put up a patch later and maintain CI pipeline for building `pandoc.wasm`.

> which would make the API dependent on the build configuration. That's not desirable.

I agree, that was just a rather dirty hack to get it working. Removing `HttpException` could remove the dependency on network related packages so that change would be very welcome. But alternatively, we could encapsulate all http client functionality in a single module (which will also encapsulate all the added CPPs), and that module would either export upstream `HttpException` or provide a stub one.

**Reply by jgm** [reply]
> But alternatively, we could encapsulate all http client functionality in a single module (which will also encapsulate all the added CPPs), and that module would either export upstream HttpException or provide a stub one.

This seems a cleaner approach.

I'm not sure what you mean "provide a stub one."  You mean duplicate the constructors on the [one from http-client](https://hackage.haskell.org/package/http-client-0.7.17/docs/Network-HTTP-Client.html#t:HttpException)?  This would also require HttpExceptionContext and Request...

Alternatively, we could just change this to a string (the result of `show`).

It looks like currently the patch requires depending on the unreleased github versions of a number of libraries.  So, one approach would be to wait until the needed changes have been released, to make the patch smaller and easier to maintain.

**Reply by TerrorJack** [reply]
> You mean duplicate the constructors on the [one from http-client](https://hackage.haskell.org/package/http-client-0.7.17/docs/Network-HTTP-Client.html#t:HttpException)?

I just meant something like `data HttpException = HttpException`. It's not intended to be compatible with the real `HttpException` anyway, but it would help minimize the changes to other parts of `pandoc` and avoid a public API change.

> So, one approach would be to wait until the needed changes have been released, to make the patch smaller and easier to maintain.

There are some patches that either the maintainer is unwilling to merge (maintenance burden), or the maintainer has ditched Haskell and archived the repos altogether. So we'll have to live with a `cabal.wasm.project` or some wasm-specific logic in `cabal.project` anyway. So I'd say there's little point in waiting for upstreaming of patches; I can try to upstream some of them concurrently as I try to add wasm build to this repo, wdyt?

**Reply by jgm** [reply]
> avoid a public API change.

Would it though?  The constructor's argument would be a *different* type that just happens to have the same name.  So the API would be different.

> So I'd say there's little point in waiting for upstreaming of patches; I can try to upstream some of them concurrently as I try to add wasm build to this repo, wdyt?

Yes, okay.

> So we'll have to live with a cabal.wasm.project or some wasm-specific logic in cabal.project anyway.

I'm not entirely sure what can be done in a cabal.project.  Is there a way to make things conditional on a build configuration?  If not, cabal.wasm.project would be needed. The problem then would be keeping this up to date; pandoc's cabal.project often changes (at least in development) when we need to use dev versions of dependencies, and this would be one more place the changes needed to be duplicated.  Or is there a way to include one cabal.project in another?

**Reply by TerrorJack** [reply]
>  Is there a way to make things conditional on a build configuration?

It is possible to put `source-repository-package` under a `if arch(wasm32)` clause so I think we don't need to worry about maintaining a separate `.project` file. An example is https://github.com/tweag/ghc-wasm-miso-examples/blob/main/cabal.project

**Reply by jgm** [reply]
If we do this I'd also like to provide `pandoc.wasm` in the Releases section.
That will make it quite easy for people to use.

**Reply by tarleb** [reply]
I'd be very interested to hear tips on how to get Lua working on wasm. Which steps would be necessary? Would we need to compile Lua separately and create a completely new interface with JS FFI to interface with it?

**Reply by TerrorJack** [reply]
Re lua support: I gave it a quick try today, and there are a couple of issues:

- lua makes use of some posix functions that are not present on wasi, the most troublesome one is `tmpnam`. This can be fixed with either assuming `/tmp` is present (and bringing in non portable assumption about the virtual fs is provided by the wasm runtime), or letting `os_tmpname` always bail out with error. I prefer the latter, though I don't know how large it might impact downstream users like pandoc filters.
- lua makes use of `setjmp`/`longjmp`. This is already supported by `wasi-sdk` by lowering to legacy eh opcodes, supported by chrome/firefox/safari, but it's not supported by wasmtime, so it hurts the portability of wasm module a bit.
- isocline does not work for wasm32-wasi and needs to be patched, you can't expect anything terminal related functionality unless you're rolling custom tty functionality outside of wasi standard using something like xterm.js

No real challenges here. I'm not including lua support in my initial pandoc.wasm just for simplicity, before we bring in lua we need to settle down on the tradeoffs, and then the implementation would not be hard.

**Reply by jgm** [reply]
This is encouraging. Because our build process already allows you to build pandoc without Lua support, we could provide both a portable wasm without Lua support and a less portable one with Lua support.

@tarleb would be able to say whether we ever use `os_tmpname`.

**Reply by tarleb** [reply]
I believe that `os.tmpname` is almost never used in third-party scripts, and definitely not in our scripts. The most common use-cases require a temporary directory, which can be generated via `pandoc.system.with_temporary_directory`, to that's what most people use. But even then, most filters that use tempfiles of any kind would require additional software and wouldn't work on wasm anyway, so it seems perfectly fine to break them and just ignore tempfiles.

**Reply by jgm** [reply]
> lua makes use of some posix functions that are not present on wasi, the most troublesome one is `tmpnam`. This can be fixed with either assuming `/tmp` is present (and bringing in non portable assumption about the virtual fs is provided by the wasm runtime), or letting `os_tmpname` always bail out with error. I prefer the latter, though I don't know how large it might impact downstream users like pandoc filters.

Is it possible to have conditional behavior? That is, check for `/tmp` in the vfs, and if it is not present, bail with error?  But I agree with @tarleb that this shouldn't be a big issue even if we error (provided `tmpnam` isn't required for the Lua interpreter itself).

> lua makes use of `setjmp/longjmp`. This is already supported by wasi-sdk by lowering to legacy eh opcodes, supported by chrome/firefox/safari, but it's not supported by wasmtime, so it hurts the portability of wasm module a bit.

Is it possible that support will be added to wasmtime?

> isocline does not work for wasm32-wasi and needs to be patched, you can't expect anything terminal related functionality unless you're rolling custom tty functionality outside of wasi standard using something like xterm.js

We don't need the Lua repl in the wasm version, I think, and we could put support for it behind a cabal flag.

**Reply by tarleb** [reply]
I believe @cscheid has been doing some experiments with Lua in wasm, too. @cscheid, do you have some additional insights on these issues?

**Reply by jgm** [reply]
In my pandoc-wasm branch I have now added a `repl` flag on pandoc-cli and pandoc-lua-engine; this is disabled for wasm builds, and that solves the problem with isocline, at least.  The next compilation error comes from hslua: ` src/Lua/Constants.hsc:107 directive const_str cannot be handled in cross-compilation mode`.  [EDIT: aha, found the `cross-compile` flag on `lua`, which fixes this!]

**Comment by jgm** [comment]
Looking at this again.  I'd love to make some changes that would make pandoc easy to compile to wasm.

For reference, here are the changes @TerrorJack had to make to the pandoc code: https://github.com/jgm/pandoc/compare/main...haskell-wasm:pandoc:wasm

@TerrorJack has the situation changed much since 2024?  Would we still need a long cabal.project with special flags for these various dependencies, and in many cases forks?  Or have we gotten to the point where most of the dependencies don't need special treatment?

**Reply by TerrorJack** [reply]
i'll revisit it after holiday season. last time i was too deep into rabbit hole of lua support that i got burnt out halfway, maybe we could settle down on upstreaming without lua first. re dependencies, i think we can live with less wasm-specific dependencies these days, we'll see once i do a second try.

**Comment by jgm** [comment]
I have put some work into this.  See the `wasm` branch in this repository for a very slimmed down version of @TerrorJack's commit that adds support for compiling to wasm.  I removed things in cabal.project that no longer seem necessary, and the new `http` build flag in pandoc.cabal allows us to avoid the more intrusive commits in the code.  (For now, Lua support is disabled.)  Unfortunately we still do seem to need forked versions of several dependencies, and these are probably now a year or more behind the current versions.

@TerrorJack, currently the playground only works for text formats (you can't produce or read a docx, for example).  If you try to produce a binary format, you get the output "Type error" -- I'm not sure where this is coming from exactly, and I don't know if it's a limitation in pandoc.wasm or just in the JS interface.  But it would be good to have an interface capable of handling binary formats.  One thought I had is that we could have the pandoc.wasm interface read and write MIME.

**Reply by jgm** [reply]
I have got a version of the playground now that (i) works with JSON-encoded options rather than command-line syntax, (ii) allows you to upload files, and (iii) allows you to specify an "output-file" (which is then added to the list of uploaded files and can be downloaded from there).  Binary formats work for input and output files.  This unlocks most of pandoc's functionality, though Lua filters are still not supported.  I'll make a demo available when I have a chance.

**Reply by jgm** [reply]
Here's a demo of my progress building on @TerrorJack 's work.
https://pandoc.org/wasm-demo/
The demo has a clunky interface; you just specify a JSON version of a defaults file, and upload any input files and auxiliary files you need.  But it shows that the core functions work.

**Reply by simonmichael** [reply]
love it!

**Reply by jgm** [reply]
Here's a demo of a user-friendly wrapper around pandoc.wasm, produced in 10 minutes by Claude: https://pandoc.org/wasm-demo/convert.html

**Comment by jgm** [comment]
@TerrorJack if you have time to have a quick look
https://github.com/jgm/pandoc/blob/wasm/cabal.project#L35-L119
I would love to get this compiling with Lua support, even if we have to make various compromises. Currently the build is failing with linker errors; I suspect something very basic is missing, but I don't understand all the pieces.  

We're using this flake to get access to ghc-wasm-meta:
https://github.com/jgm/pandoc/blob/wasm/flake.nix#L1-L80

To trigger building with Lua it's sufficient to change `-lua` to `+lua` under pandoc-cli, and then `make pandoc.wasm`.

**Reply by TerrorJack** [reply]
what's the detailed linker error and the exact build command?

**Reply by jgm** [reply]
```
wasm-ld: warning: function signature mismatch: lua_version
>>> defined as (i32) -> i32 in /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(Primary.o)
>>> defined as (i32) -> f64 in /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(lapi.o)

wasm-ld: warning: function signature mismatch: hslua_get_caching_table
>>> defined as (i32, i32) -> i32 in /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/hsl-bjctrnttn-2.5.0-dfeddfbc/lib/libHShsl-bjctrnttn-2.5.0-dfeddfbc.a(hsllist.o)
>>> defined as (i32, i32) -> void in /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/hsl-bjctrnttn-2.5.0-dfeddfbc/lib/libHShsl-bjctrnttn-2.5.0-dfeddfbc.a(hslobj.o)

wasm-ld: error: /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(ldo.o): undefined symbol: __wasm_longjmp
wasm-ld: error: /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(ldo.o): undefined symbol: __wasm_setjmp
wasm-ld: error: /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(ldo.o): undefined symbol: __c_longjmp
wasm-ld: error: /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(ldo.o): undefined symbol: __wasm_setjmp_test
wasm-ld: error: /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(ldo.o): undefined symbol: __wasm_longjmp
wasm-ld: error: /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(loslib.o): undefined symbol: clock
wasm-ld: error: /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(loslib.o): undefined symbol: system
wasm-ld: error: /Users/jgm/.ghc-wasm/.cabal/store/ghc-9.14.1.20251219-944e/l-2.3.4-928c1010/lib/libHSl-2.3.4-928c1010.a(ltablib.o): undefined symbol: clock
wasm32-wasi-clang: error: linker command failed with exit code 1 (use -v to see invocation)
wasm32-wasi-ghc: `wasm32-wasi-clang' failed in phase `Linker'. (Exit code: 1)
HasCallStack backtrace:
  collectBacktraces, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:92:13 in ghc-internal:GHC.Internal.Exception
  toExceptionWithBacktrace, called at libraries/ghc-internal/src/GHC/Internal/IO.hs:260:11 in ghc-internal:GHC.Internal.IO
  throwIO, called at libraries/exceptions/src/Control/Monad/Catch.hs:371:12 in exceptions-0.10.7-b950:Control.Monad.Catch
  throwM, called at libraries/exceptions/src/Control/Monad/Catch.hs:860:84 in exceptions-0.10.7-b950:Control.Monad.Catch
  onException, called at compiler/GHC/Driver/MakeAction.hs:235:23 in ghc-9.14.1.20251219-944e:GHC.Driver.MakeAction


Error: [Cabal-7125]
Failed to build exe:pandoc from pandoc-cli-3.8.3.
```
For ghc wasm dependencies I'm using `gitlab:haskell-wasm/ghc-wasm-meta?host=gitlab.haskell.org` via nix flake.

**Reply by jgm** [reply]
Build command: in the `wasm` branch: `wasm32-wasi-cabal build pandoc-cli` after changing `-lua` to `+lua` in `cabal.project`'s stanza for `pandoc-cli`.

**Reply by kamelkace** [reply]
Apologies for cutting into your existing conversation, but in case it's helpful, I was able to get the build a little further along by setting the linker flags on the `pandoc-cli` package instead of `lua`:
```
diff --git a/cabal.project b/cabal.project
index 0a44438a1..3d9fc2ffc 100644
--- a/cabal.project
+++ b/cabal.project
@@ -58,7 +58,8 @@ if arch(wasm32)
     flags: +embed_data_files -http
 
   package pandoc-cli
-    flags: -lua -server
+    flags: +lua -server
+    ld-options: -lwasi-emulated-process-clocks -lwasi-emulated-signal -lsetjmp
```
The build still fails on `(loslib.o): undefined symbol: system`; however, I was able to get past that point in a test project by patching out some code in the vendor'd copy of Lua in hslua. I'm still figuring out how to actually run the code from that test -- wasmtime needs to be built with support for legacy exceptions in order to run the wasm binaries that're generated -- but once that's available I'll send over a link.

**Reply by jgm** [reply]
Oh, yes, of course that's where the linker flags need to go! Thank you, very helpful.

**Reply by kamelkace** [reply]
As noted, [here's a repo](https://github.com/kamelkace/minimal-hslua-wasm-example) demonstrating all the flags + patches that I needed to add to get [a basic hslua "Hello World"](https://github.com/kamelkace/minimal-hslua-wasm-example/blob/main/app/Main.hs) running. My knowledge of haskell is quite poor (although it's quite an interesting language!), and a bunch of this was written via trial-and-error + LLM slop sifting, so there could easily be some unnecessary bits that stuck around -- just as an FYI.

**Reply by jgm** [reply]
I was able to compile pandoc.wasm with Lua filter support -- tested with a simple filter and it worked!

(I ended up doing something similar to what you have, with a custom version of hslua.  But your approach is better. I didn't know you could use post-checkout-command to apply patches in a cabal.project !   I'll update my branch to use this approach.

**Reply by jgm** [reply]
However, I now get an error on the `wasm-opt` step:
`unexpected expr type
UNREACHABLE executed at /Users/runner/work/binaryen/binaryen/src/passes/Flatten.cpp:231!`.
I can successfully build an unoptimized version of pandoc.wasm.

**Reply by kamelkace** [reply]
It looks like an error handler somewhere in the bytecode is causing the optimizer's flatten stage to fail. This is what I'm seeing in rr ([source](https://github.com/WebAssembly/binaryen/blob/660166319b6cdb255b5dbc3d2a3ef3e9c0a00ad3/src/passes/Flatten.cpp#L231), albeit, a few lines away since `curr` is optimized out by line 231):
```
(rr) call curr->dump()
(try_table (catch $tag$0 $block2)
 (unreachable)
)
```
From poking around in the debug output of `wasm-opt`/binaryen, I don't think there's a preprocessing stage that is creating this in the bytecode, so it's probably either (1) a bug in binaryen's handling of exception handlers, or (2) a bug in the WASM bytecode generation of something else in the stack (or both, I suppose). Diagnosing _that_ requires more context than I have, but adding `--strip-eh` (per [the docs](https://github.com/WebAssembly/binaryen/blob/main/src/passes/StripEH.cpp#L19), thrown exceptions will now crash) seems to get the optimizer working. Not sure if that'll break something else, however.
```
# example
wasm-opt -d --low-memory-unused --converge --gufa --strip-eh --flatten --rereloop -Oz $(wasm32-wasi-cabal list-bin pandoc-cli | tail -1) -o pandoc.wasm
```

**Reply by jgm** [reply]
To keep things simple for now, I think I'll just strip it down to `-Oz`. That gives me a 56M binary instead of 54M with no error.

**Reply by jgm** [reply]
Added a Lua filters example: https://pandoc.org/wasm-demo/
It is working!

**Reply by simonmichael** [reply]
Congrats! The demo is really featureful and impressive and the examples are very helpful. A lovely addition would be showing the equivalent command line.

**Reply by tarleb** [reply]
Amazing! :rocket:

**Comment by simonmichael** [comment]
Is https://pandoc.org/wasm-demo broken today ? It spins forever at "Loading pandoc.wasm..." for me.

**Reply by simonmichael** [reply]
Just me I think; switching to my normal wifi fixed it.

**Reply by jgm** [reply]
It was indeed broken for a short time.

**Comment by tarleb** [comment]
I'm forwarding three of the responses that the announcement post on Mastodon received:

---

> If I can give a suggestion: don't block the whole UI while loading the wasm, but just the "convert" button, so that people can pick their document and options in parallel to the code loading

---

> \#pandoc is one of the world wonders! Thank you!

---

> Pandoc is about as close as civilisation has come to magic so far. (ffmpeg is also close.)

**Reply by jgm** [reply]
Good suggestion on the blocking.  I will look into that.  [DONE!]

**Reply by tarleb** [reply]
One more:

> [@pandoc](https://fosstodon.org/@pandoc) incredible work! I'll definitely be using this extensively :)
>
> May I report a bug? It seems like files with names in Arabic (or other Unicode non-ASCII characters) gives this error: `ERROR: Unexpected end-of-input while parsing string literal` 
>
> Changing the file name to ASCII characters fixes the problem.
>
> Thanks again for your work 🙏

**Reply by jgm** [reply]
Added https://github.com/jgm/pandoc/issues/11447 to  track this

**Reply by jgm** [reply]
I had Claude diagnose it; the problem, it suggests, is that the JavaScript string length is used to allocate space for the UTF-8 encoded string, which leads to not enough space being allocated when you have code points that need 3 or 4 bytes.

**Reply by jgm** [reply]
OK, it seems to be fixed now.  Let me know if there are more issues.

**Comment by catamaican** [comment]
For me, it's was not working in Firefox 32bits (Failed to load pandoc: InternalError: out of memory), https://pandoc.org/wasm-demo/ or "/app". On 64bit it's working. I had no trouble on MS Edge
